User interface acceptance criteria: Submit order (with prerequisites) (#1181)
vegaprotocol/specs: user-interface/0003-WTXN-submit_vega_transaction.md

# Submit Vega transaction

> A dapp sends a transaction to a wallet, that wallet then broadcasts the transaction to a network.  Therefore the following is broken up into two steps. The transaction could fail at either. Generally: Once the transaction has gone to the network a use can use block explorer to track the transaction, but some tracking in Dapp or wallet will help.

Reviewer: ```suggestion
A dapp sends a transaction to a wallet, that wallet then broadcasts the transaction to a network.  Therefore the following is broken up into two steps. The transaction could fail at either. Generally: Once the transaction has gone to the network a user can use block explorer to track the transaction, but some tracking in Dapp or wallet will help.
```

When submitting a Vega transaction of any kind, I...

## Track transaction to wallet

if not connected to a Vega wallet:

- **must** be told that I am not connected and given the [option to connect](0012-WCON-connect_vega_wallet.md) <a name="0003-WTXN-001" href="#0003-WTXN-001">0003-WTXN-001</a>

if transaction not auto approved by wallet:

- **must** see a prompt to check connected vega wallet to approve transaction <a name="0003-WTXN-002" href="#0003-WTXN-002">0003-WTXN-002</a>

if order is approved by wallet:

> - **must** see A hash/transaction ID <a name="0003-WTXN-003" href="#0003-WTXN-003">0003-WTXN-003</a>

Reviewer: as well as a label telling you what it is, with naming that is consistent across dapps

Author: explanation added in https://github.com/vegaprotocol/specs/pull/1181/commits/8fa3efed0a99c1bf2ac27c3eeb9775b70ffe1760
```suggestion
- **must** see A [transaction hash](7001-DATA-data_display.md#transaction-hash) <a name="0003-WTXN-003" href="#0003-WTXN-003">0003-WTXN-003</a>
```

- **must** see the public key that this transaction was submitted for <a name="0003-WTXN-004" href="#0003-WTXN-004">0003-WTXN-004</a>
- **should** see the alias for the key that submitted this transaction <a name="0003-WTXN-005" href="#0003-WTXN-005">0003-WTXN-005</a>

> - **could** see a prompt to see this app to [auto approve](0001-WALL-wallet.md#approving-transactions) in wallet app <a name="0003-WTXN-006" href="#0003-WTXN-006">0003-WTXN-006</a>

Reviewer 2: ```suggestion
- **could** see a prompt to set this app to [auto approve](0001-WALL-wallet.md#approving-transactions) in wallet app <a name="0003-WTXN-006" href="#0003-WTXN-006">0003-WTXN-006</a>
```

if transaction is rejected by wallet:

- **must** see that the order was rejected by the connected wallet <a name="0003-WTXN-007" href="#0003-WTXN-007">0003-WTXN-007</a>

if the wallet does not respond:

- **must** be able to cancel attempt to submit transaction <a name="0003-WTXN-008" href="#0003-WTXN-008">0003-WTXN-008</a>

if the wallet highlights an issue with the transaction:

- **must** show that the transaction was marked as invalid by the wallet and not broadcast <a name="0003-WTXN-009" href="#0003-WTXN-009">0003-WTXN-009</a>
- **should** see the error returned highlighted in context of the form that submitted the transaction in Dapp <a name="0003-WTXN-010" href="#0003-WTXN-010">0003-WTXN-010</a>
- **must** show error returned by wallet <a name="0003-WTXN-011" href="#0003-WTXN-011">0003-WTXN-011</a>

## Track transaction on network 

- **must** see a link to that transaction in a block explorer for the appropriate network <a name="0003-WTXN-012" href="#0003-WTXN-012">0003-WTXN-012</a>

> - **must** see the transaction status - TODO Document these <a name="0003-WTXN-013" href="#0003-WTXN-013">0003-WTXN-013</a>

Reviewer 2: ```suggestion
- **must** see the transaction status - TODO Document these <a name="0003-WTXN-013" href="#0003-WTXN-013">0003-WTXN-013</a>
- **must** see the network the transaction was broadcast to <a name="0003-WTXN-XXX" href="#0003-WTXN-XXX">0003-WTXN-XXX</a>
```

- **must** see the block the transaction was processed in <a name="0003-WTXN-014" href="#0003-WTXN-014">0003-WTXN-014</a>
- **must** show the node the transaction was broadcast to <a name="0003-WTXN-015" href="#0003-WTXN-015">0003-WTXN-015</a>
- **must** see the validator that processed the block the transaction was processed in <a name="0003-WTXN-016" href="#0003-WTXN-016">0003-WTXN-016</a>
- **must** see the content of the transaction as seen by the network <a name="0003-WTXN-017" href="#0003-WTXN-017">0003-WTXN-017</a>

> ... so I am aware of the transactions status interns of being sent to and processed by the network

Reviewer: ```suggestion
... so I am aware of the transactions status of the transactions my wallet is sending and that are being processed by the network
```

# Submit Vega transaction

A dapp sends a transaction to a wallet, that wallet then broadcasts the transaction to a network.  Therefore the following is broken up into two steps. The transaction could fail at either. Generally: Once the transaction has gone to the network a user can use block explorer to track the transaction, but some tracking in Dapp or wallet will help.

When submitting a Vega transaction of any kind, I...

## Track transaction to wallet

if not connected to a Vega wallet:

> - **must** be told that I am not connected and given the [option to connect](0012-WCON-connect_vega_wallet.md) (<a name="0003-WTXN-001" href="#0003-WTXN-001">0003-WTXN-001</a>)

Reviewer 3: You wouldn't be able to submit a transaction if you weren't connected would you? Perhaps that's what this is saying, but maybe it could say 'Be prevented from submitting the transaction, with it made clear that I need to connect my wallet first'. Or just broken out from the 'when submitting a transaction' section.

Author: ```suggestion
- **must** be told that I am not connected, so can not submit a transaction, and given the [option to connect](0012-WCON-connect_vega_wallet.md). (Note: this may have happened if the wallet has become disconnected without the user knowing) (<a name="0003-WTXN-001" href="#0003-WTXN-001">0003-WTXN-001</a>)
```

> if transaction not auto approved by wallet:

Reviewer 3: Probably a given, but we don't actually state that if auto-approve isn't on the user should see an approve/reject dialog.

Author: I think this covers it?

```Quote
if transaction not auto approved by wallet:

- **must** see a prompt to check connected vega wallet to approve transaction (<a name="0003-WTXN-002" href="#0003-WTXN-002">0003-WTXN-002</a>)
```

Reviewer 3: That's a message on the app side. Feels pressing approve/reject on the wallet side is fundamental enough to have its own AC.

Author: It does. 
0001-WALL-024
https://github.com/vegaprotocol/specs/blob/user-interface-SORD-only/user-interface/0001-WALL-wallet.md#approving-transactions

Or am I missing something?

> - **must** see a prompt to check connected vega wallet to approve transaction (<a name="0003-WTXN-002" href="#0003-WTXN-002">0003-WTXN-002</a>)

Reviewer 3: Should we also mention what transaction info both the dapp and the wallet should show at this stage?

Author: The wallet file `0001-WALL-wallet.md` **Approving transactions** heading. Shows that it needs to show the transaction detail.

Added the following to say that a dapp could show the user what has been sent to the wallet

```suggestion
- **must** see a prompt to check connected vega wallet to approve transaction (<a name="0003-WTXN-002" href="#0003-WTXN-002">0003-WTXN-002</a>)
- **could** see the transaction details that has been passed to the wallet for broadcast (<a name="0003-WTXN-019" href="#0003-WTXN-019">0003-WTXN-019</a>)
```

> if order is approved by wallet:

Reviewer 3: Clarify whether we are we talking about showing these in the dapp or in the wallet here

Author: I think I'm going add something to the Read me that says that everything except `0001-WALL-wallet` is about what you see in dApp. Would that resolve your comment @PeteWilliams ?

Author: see https://github.com/vegaprotocol/specs/pull/1181/commits/229d859b2a236095a13b406140a9631e155f844e
